### PR TITLE
Coordinated quota

### PR DESCRIPTION
**Pull request by efirs**
Quota package made composable and broken down into:                              
  * Node quota (total limits one node can handle)                                
  * Storage quota (maximum namespace size quota)                                 
  * Per cluster namespace quota, with Datadog backend                            
                                                                                 
Storage quota, size metrics and namespace rate updates                           
moved to background threads.                                                     
                                                                                 
Observability code is generalized and Datadog backend                            
is used by both O11y and Quota.                                                  
                                                                                 
Unrelated tests fixes

### CODE REVIEW

File: server/config/options.go [hunk 1]
```diff
@@ -157,17 +162,42 @@ var DefaultConfig = Config{
...
-		ReadThroughputLimit:       10000000,    // bytes per second
-		DataSizeLimit:             10000000000, // bytes
-		LimitUpdateInterval:       5,           // seconds
-		TenantSizeRefreshInterval: 60,          // seconds
-		AllTenantsRefreshInternal: 300,         // seconds
+		Node: LimitsConfig{
+			Enabled: false,
+
+			ReadRateLimit:        1000,     // read requests per second per namespace
+			ReadThroughputLimit:  10000000, // read bytes per second
```
Comment (efirs) [inline server/config/options.go:169]: If Throughput is not set then it deduces from rate and unit size.

File: api/server/v1/tx.go [hunk 2]
```diff
@@ -21,24 +21,70 @@ import (
...
+		return true
+	case describeCollection, describeDatabase:
+		return true
+	default:
+		return false
+	}
+}
+
+func isWrite(name string) bool {
+	return !isRead(name)
```
Comment (JigarJoshi) [inline api/server/v1/tx.go:70]: we classified requests in three categories for metrics query api. read, write and metadata. https://github.com/tigrisdata/tigris/blob/main/server/services/v1/observability.go#L213 from quota perspective what you did makes sense but just bringing it up here if we should restrict the usage of N number of db creation for example.

File: api/server/v1/tx.go [hunk 3]
```diff
@@ -21,24 +21,70 @@ import (
...
+	case listCollections, listDatabases:
+		return true
+	case describeCollection, describeDatabase:
+		return true
+	default:
+		return false
+	}
+}
+
+func isWrite(name string) bool {
```
Comment (JigarJoshi) [inline api/server/v1/tx.go:69]: probably we can move these methods in `server/request/request.go` in `RequestMetadata`. wdyt?

File: api/server/v1/tx.go (hunk 2 above)
Comment (ovaistariq) [reply api/server/v1/tx.go:70]: Yeah we don't count list database, list collection, etc as read queries when displaying in the web console

File: server/quota/node.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+// This limiter protects the node from overloading,
+// It limits the total number of request and throughput
+// handled by the node, by all the namespaces.
+
+type node struct {
+	state *State
+	cfg   *config.QuotaConfig
+}
+
+func (c *node) Allow(_ context.Context, namespace string, size int, isWrite bool) (err error) {
```
Comment (himank) [inline server/quota/node.go:19]: not seeing much benefit of named return value here, same applies to below method.

File: server/quota/namespace.go [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+	"go.uber.org/atomic"
+	"golang.org/x/time/rate"
+)
+
+const (
+	rateHysteresis       = 10         // ±10 rps per instance wouldn't cause rate regulation
+	throughputHysteresis = 100 * 1024 // ±100Kb per instance wouldn't cause throughput regulation
+
+	rateIncrement       = 5     // when rate adjustment is needed increment current rate by ± this value
+	throughputIncrement = 10000 // when throughput adjustment is needed increment by ± this value
```
Comment (himank) [inline server/quota/namespace.go:20]: don't we need to adjust it by (+-)10% of the rate?

File: server/quota/quota.go [hunk 6]
```diff
@@ -17,35 +17,32 @@ package quota
...
-	"github.com/tigrisdata/tigris/server/transaction"
 	"go.uber.org/atomic"
-	"golang.org/x/time/rate"
 )
 
 var (
-	ErrRateExceeded        = errors.ResourceExhausted("request rate limit exceeded")
-	ErrThroughputExceeded  = errors.ResourceExhausted("request throughput limit exceeded")
-	ErrStorageSizeExceeded = errors.ResourceExhausted("data size limit exceeded")
+	ErrReadRateExceeded        = errors.ResourceExhausted("request read rate limit exceeded")
```
Comment (himank) [inline server/quota/quota.go:30]: why not just use `api.Errorf(api.Code_ResourceExhausted...)`?

**Comment by himank** [comment]
Overall very cool!

File: server/config/options.go [hunk 7]
```diff
@@ -157,17 +162,42 @@ var DefaultConfig = Config{
...
+		Storage: StorageLimitsConfig{
+			Enabled:         false,
+			DataSizeLimit:   100 * 1024 * 1024,
+			RefreshInternal: 60 * time.Second,
+		},
 	},
 	Observability: ObservabilityConfig{
-		Enabled: false,
+		Provider:    "datadog",
+		ProviderUrl: "https://us3.datadoghq.com",
```
Comment (pboros) [inline server/config/options.go:199]: Should this be default configuration?

File: api/server/v1/tx.go (hunk 2 above)
Comment (efirs) [reply api/server/v1/tx.go:70]: Right, there is discrepancy, but we need to enforce quota even for metadata requests, so we classify them as reads and writes here.

File: server/config/options.go [hunk 8]
```diff
@@ -157,17 +162,42 @@ var DefaultConfig = Config{
...
-		WriteThroughputLimit:      10000000,    // bytes per second
-		ReadThroughputLimit:       10000000,    // bytes per second
-		DataSizeLimit:             10000000000, // bytes
-		LimitUpdateInterval:       5,           // seconds
-		TenantSizeRefreshInterval: 60,          // seconds
-		AllTenantsRefreshInternal: 300,         // seconds
+		Node: LimitsConfig{
+			Enabled: false,
+
+			ReadRateLimit:        1000,     // read requests per second per namespace
```
[the remark this answers is not in the document]
Comment (efirs) [reply server/config/options.go:168]: This is per node config. I'll adjust it after testing.

File: server/config/options.go [hunk 9]
```diff
@@ -187,17 +217,61 @@ type SearchConfig struct {
...
+func (l *LimitsConfig) ReadThroughput() int64 {
+	if l.ReadThroughputLimit != 0 {
+		return l.ReadThroughputLimit
+	}
+	return int64(ReadUnitSize) * int64(l.ReadRateLimit)
+}
+
+type NamespaceLimitsConfig struct {
+	Enabled         bool
+	Default         LimitsConfig            // default per namespace limit
```
[the remark this answers is not in the document]
Comment (efirs) [reply server/config/options.go:245]: Default is default per namespace per cluster config.
Node is maximum per namespace per node config, single namespace cannot use more than this on one node.

File: server/quota/quota.go (hunk 6 above)
Comment (efirs) [reply server/quota/quota.go:30]: `errors.ResourceExhausted(` is wrapper of `api.Errorf(api.Code_ResourceExhausted...`

File: server/config/options.go (hunk 7 above)
Comment (efirs) [reply server/config/options.go:199]: > Should this be default configuration?

Do you mean have empty default instead?

> Maybe if there is no observability configured, only do local limits.

Good point. Changed it that way.

File: server/metrics/datadog.go [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	if hResp.StatusCode == http.StatusTooManyRequests {
+		log.Warn().Str(rateLimitLimit, hResp.Header.Get(rateLimitLimit)).
+			Str(rateLimitPeriod, hResp.Header.Get(rateLimitPeriod)).
+			Str(rateLimitRemaining, hResp.Header.Get(rateLimitRemaining)).
+			Str(rateLimitReset, hResp.Header.Get(rateLimitReset)).
+			Str(rateLimitName, hResp.Header.Get(rateLimitName)).
+			Msgf("Datadog rate-limit hit")
+		return nil, errors.ResourceExhausted("Failed to get query metrics: reason = rate-limited, reason = %s", resp.GetError())
```
[the remark this answers is not in the document]
Comment (efirs) [reply server/metrics/datadog.go:56]: That what `errors.ResourceExhausted(` returns.

File: server/config/options.go [hunk 11]
```diff
@@ -157,17 +162,42 @@ var DefaultConfig = Config{
...
 	Quota: QuotaConfig{
-		Enabled:                   false,
-		RateLimit:                 1000,        // requests per second both reads and writes
-		WriteThroughputLimit:      10000000,    // bytes per second
-		ReadThroughputLimit:       10000000,    // bytes per second
-		DataSizeLimit:             10000000000, // bytes
-		LimitUpdateInterval:       5,           // seconds
-		TenantSizeRefreshInterval: 60,          // seconds
-		AllTenantsRefreshInternal: 300,         // seconds
+		Node: LimitsConfig{
```
[the remark this answers is not in the document]
Comment (efirs) [reply server/config/options.go:165]: Right. outside is across all namespaces, absolute maximum node allowed to handle.

[the remark this answers is not in the document]
Comment (efirs) [reply server/config/options.go:165]: Inside is maximum per namespace per node.

File: api/server/v1/tx.go (hunk 2 above)
Comment (efirs) [reply api/server/v1/tx.go:70]: Fixed. There are now only two categories for the quota metrics request. Check: `FormDatadogQueryNoMeta`

File: server/quota/namespace.go (hunk 5 above)
Comment (efirs) [reply server/quota/namespace.go]: This is good catch. Fixed.

File: server/config/options.go [hunk 12]
```diff
@@ -275,17 +280,38 @@ var DefaultConfig = Config{
...
+
+			ReadUnits:  4000, // read requests per second per namespace
+			WriteUnits: 1000, // write requests per second
+		},
+		Namespace: NamespaceLimitsConfig{
+			// Per cluster limits for single namespace
+			Default: LimitsConfig{
+				Enabled: false,
+
+				ReadUnits:  400, // read requests per second per namespace
```
Comment (himank) [inline server/config/options.go:295]: make it 100.

File: server/config/options.go [hunk 13]
```diff
@@ -275,17 +280,38 @@ var DefaultConfig = Config{
...
+				WriteUnits: 100, // write requests per second
```
Comment (himank) [inline server/config/options.go:296]: and write can be 25.

File: server/quota/namespace.go [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+	}
+
+	return setNodeLimit
+}
+
+func (i *namespace) updateLimitsLow(set int64, cur int64, maxInstRate int64, maxNodeRate int) int64 {
+	return calcLimit(set, int64(maxNodeRate), cur, maxInstRate, rateHysteresis, rateIncrement)
+}
+
+func (i *namespace) updateLimits(ns string, is *instanceState, curRead int64, curWrite int64) {
```
Comment (himank) [inline server/quota/namespace.go:151]: Probably have some metrics for the new limits, basically what are the new limits for a namespace.

File: server/quota/quota.go [hunk 15]
```diff
@@ -50,215 +49,102 @@ type State struct {
...
-		defer s.TenantSizeLock.Unlock()
-
-		s.TenantSizeUpdateAt.Store(currentTimeStamp)
-		metrics.UpdateNameSpaceSizeMetrics(namespace, sz)
-		m.updateTenantSize(ctx, namespace)
+// Allow checks read, write rates and storage size limits for the namespace
+// and returns error if at least one of them is exceeded.
+func Allow(ctx context.Context, namespace string, size int, isWrite bool) error {
+	for _, q := range mgr.quota {
+		if err := q.Allow(ctx, namespace, size, isWrite); err != nil {
```
Comment (himank) [inline server/quota/quota.go:130]: Here best to have metrics if we are not allowing and which rate-limiter is kicking in.

**Review by himank — DISMISSED** [review]

**Review by himank — APPROVED** [review]

**Pull request merged**